src/Bicep.Core.UnitTests/TypeSystem/FunctionResolverTests.cs

```diff
@@ -174,6 +174,421 @@ public void SplitReturnsNonEmtpyArrayOfStrings()
+                new FunctionArgumentSyntax[] { new(TestSyntaxFactory.CreateVariableAccess("foo")), new(TestSyntaxFactory.CreateVariableAccess("bar")) })
+                .Type;
+
+            returnType.Should().BeAssignableTo<ArrayType>();
+            returnType.As<ArrayType>().MaxLength.Should().NotBeNull();
+            returnType.As<ArrayType>().MaxLength.Should().Be(21);
+        }
+
+        [TestMethod]
+        public void ConcatDoesNotDeriveMaxLengthOfReturnTypeIfAnyInputLacksAMaxLength()
+        {
+            var returnType = EvaluateFunction("concat",
+                new List<TypeSymbol> { TypeFactory.CreateArrayType(LanguageConstants.String, maxLength: 10), LanguageConstants.Array, TypeFactory.CreateArrayType(LanguageConstants.String, maxLength: 11) },
+                new FunctionArgumentSyntax[] { new(TestSyntaxFactory.CreateVariableAccess("foo")), new(TestSyntaxFactory.CreateVariableAccess("bar")), new(TestSyntaxFactory.CreateVariableAccess("baz")) })
+                .Type;
+
+            returnType.Should().BeAssignableTo<ArrayType>();
+            returnType.As<ArrayType>().MaxLength.Should().BeNull();
+        }
+
+        [TestMethod]
+        public void ConcatConcatenatesTuples()
+        {
+            var returnType = EvaluateFunction("concat",
+                new List<TypeSymbol>
+                {
+                    new TupleType(ImmutableArray.Create<ITypeReference>(LanguageConstants.String, LanguageConstants.Int), default),
+                    new TupleType(ImmutableArray.Create<ITypeReference>(LanguageConstants.Bool), default),
+                    new TupleType(ImmutableArray.Create<ITypeReference>(TypeFactory.CreateStringLiteralType("abc"), TypeFactory.CreateIntegerLiteralType(123)), default),
+                },
+                new FunctionArgumentSyntax[] { new(TestSyntaxFactory.CreateVariableAccess("foo")), new(TestSyntaxFactory.CreateVariableAccess("bar")), new(TestSyntaxFactory.CreateVariableAccess("baz")) })
+                .Type;
+
+            returnType.Should().BeAssignableTo<TupleType>();
+            returnType.As<TupleType>().Items.Should()
+                .ContainInOrder(LanguageConstants.String, LanguageConstants.Int, LanguageConstants.Bool, TypeFactory.CreateStringLiteralType("abc"), TypeFactory.CreateIntegerLiteralType(123));
+        }
+
+        [TestMethod]
+        public void SkipDerivesMinAndMaxLengthOfReturnType()
+        {
+            var returnType = EvaluateFunction("skip",
+                new List<TypeSymbol> { TypeFactory.CreateArrayType(LanguageConstants.String, minLength: 10, maxLength: 20), TypeFactory.CreateIntegerLiteralType(9) },
+                new FunctionArgumentSyntax[] { new(TestSyntaxFactory.CreateVariableAccess("foo")), new(TestSyntaxFactory.CreateVariableAccess("bar")) })
+                .Type;
+
+            returnType.Should().BeAssignableTo<ArrayType>();
```

Comment:
Here's a `FluentAssertion` trick to avoid casting `returnType` manually below:
```
var arrayType = returnType.Should().BeAssignableTo<ArrayType>().Subject;
arrayType.Item.Should().Be(LanguageConstants.String)
...
```